### DISCUSSION TITLE
Unexpected GiB in unit calculation

### DISCUSSION THREAD

**Discussion by zcohan** [josdejong/mathjs#2764] [Ideas]
I'm seeing GiB in some unit calculations, when I expect GB.

There seems to be a threshold point. Here's an example:

5 minutes * 54.74 MB/second = 16.422 GB
5 minutes * 54.75 MB/second = 15.296973288059 GiB

**Comment by josdejong** [comment]
That's an interesting one, thanks for reporting @zcohan .

Reason is both prefixes `G` and `Gi` are defined in the same list, and there is currently no distinction at to prevent switching from one to the other, see code:

https://github.com/josdejong/mathjs/blob/a514b8cb23b4839fbc82418dc00d97936c31feee/src/type/unit/Unit.js#L1338-L1356

We somehow need to distinguish the two types of prefixes. I don't have an idea right away what would be a handy approach though, @ericman314 do you have any ideas?

**Comment by ericman314** [comment]
Right, the prefix is chosen to keep the rendered value "not too big" and "not too small". I don't know of a quick fix for this, it'll take a little more thought.

**Comment by josdejong** [comment]
If anyone is interested in picking this issue up please let me know.

**Comment by jackcmason** [comment]
I would be interested in fixing this issue

**Comment by harrysarson** [comment]
Is what we need to do is similar to choosing between feet and meters for distances?

**Comment by josdejong** [comment]
Thanks for your offer @jackcmason, some help would be great. I think we first need to come up with an idea on how to solve this issue. Do you have any suggestion?

**Comment by josdejong** [comment]
@harrysarson I think it's a similar to not wanting to mix short prefixes (like `cm`) with long prefixes (like `centimeter`)

**Comment by harrysarson** [comment]
However, `1 cm` and `1 centimeter` are exactly the same length - this is not true for `1 Gb` and `1 GiB`

**Comment by josdejong** [comment]
:) yes true, I just mean they are somehow similar but should not get mixed together.

**Comment by ericman314** [comment]
The problem is, how do we decide what the user wants to see? It is a combination of choosing a prefix to keep the magnitude of the value reasonable, choosing the correct format of the prefix (long or short), and the correct unit system, SI or something else. All we have to go on is options the user configures, and what the user has used in the past.

Here's one possible approach. We could make several highly customizable options which explicitly define what units and prefixes the user will see. Things like GB or GiB. Then, to achieve the "automagic" behavior that math.js currently tries to do, the options would be selected automatically as the user is interacting with a math.js instance. That would go a long way to eliminating the mystery of what's going on behind the scenes in the unit formatter, since the user could just query the current config to see why a unit was formatted the way it was. For most users that are happy with the way things work now, it would still work the same, but for the few that need to control how units get formatted they could do so.

**Comment by jackcmason** [comment]
I think that bits should be restricted to SI or IEC prefixes when parsed or by the `_bestPrefix` method based on whether the input value is an SI or IEC unit. The user should convert between the two prefixes with a conversion method between the SI or IEC units to be called by the user explicitly.

**Comment by josdejong** [comment]
@ericman314 I like your idea, I think we should en up with putting all current magic in configurable options which have a good as possible default settings. That will empower users that want total control.

@jackcmason would be nice as a start, in many (or some?) cases we can know the typeof prefix from the input (like `KiB` or `KB`). Where we don't know it from the inputs, we should use for example EIC by default but (on long term) make this configurable by passing options to the format function.

Would be nice if we could make a first small step in the right direction splitting these prefixes, and as a second step work towards a configurable solution.

**Comment by josdejong** [comment]
@jackcmason you made a start with implementing this feature in #1390, do you feel like working on the next steps maybe?
